hypervisor/src/kvm/mod.rs

```diff
@@ -714,6 +714,56 @@ impl cpu::Vcpu for KvmVcpu {
     }
     #[cfg(target_arch = "x86_64")]
     ///
+    /// X86 specific call to check the consistency of CPUID registers.
+    ///
+    /// The function returns `true` (a.ka. "consistent"), when the `input_cpuid` (e.g. as a part of the 'saved CPU state'
+    /// from the source VM of the live-migration) is a subset of the current CPUID (e.g. built with KVM ioctls for the
+    /// destination VM of the live-migration).
+    fn check_cpuid_consistency(&self, input_cpuid: &CpuId) -> cpu::Result<bool> {
+        let mut current_cpuid = self.get_cpuid2(kvm_bindings::KVM_MAX_CPUID_ENTRIES)?;
+        current_cpuid.retain(|c| *c != Default::default());
+        let mut input_cpuid = input_cpuid.clone();
+        input_cpuid.retain(|c| *c != Default::default());
+
+        let mut consistent = true;
+        for input_entry in input_cpuid.as_slice().iter() {
+            let mut entry_found = false;
+            let mut warned = false;
+            for current_entry in current_cpuid.as_slice().iter() {
+                if current_entry.function == input_entry.function
+                    && current_entry.index == input_entry.index
+                {
+                    if current_entry == input_entry {
+                        entry_found = true;
+                        break;
+                    } else {
+                        warned = true;
```

Comment:
I think this algorithm needs to be much smarter. What we care about is if the destination set of features (as described by the CPUID) is the same or a superset of those at the source. Or to put it the other way. We need to reject a restore if the destination is lacking a feature compared to the source.

CPUID is a lot more than just features so you need to actually interpret the bits. It shouldn't matter if the model number / CPU name changes. Nor does the cache arrangement matter that much either. What were want to check for is if the destination had `avx2` then the source must also advertise `avx2`.

Reply:
> What were want to check for is if the destination had `avx2` then the source must also advertise `avx2`.

Is this something that could be useful to have configured? Some workloads may not use additional CPU features. My memory of how what AWS/Azure allow for migrations isn't very good so I don't know if it is something that the cloud providers support at some level or not.

Reply:
Good question, and let's discuss it more. If we want it, we can do it as an improvement in the future.